src/rpc/methods/eth/filter/mod.rs

```diff
@@ -0,0 +1,263 @@
+        let tipset_filter_manager = Some(TipSetFilterManager::new(max_filter_results));
+
+        Self {
+            filter_store,
+            max_filter_height_range,
+            event_filter_manager,
+            tipset_filter_manager,
+        }
+    }
+
+    pub fn eth_new_filter(
+        &self,
+        filter_spec: &EthFilterSpec,
+        chain_height: i64,
+    ) -> Result<FilterID, FilterError> {
+        if self.filter_store.is_none() || self.event_filter_manager.is_none() {
+            return Err(FilterError::NotSupported);
+        }
+
+        let pf = filter_spec
+            .parse_eth_filter_spec(chain_height, self.max_filter_height_range)
+            .map_err(FilterError::ParsingError)?;
+
+        let f = self
+            .event_filter_manager
+            .as_ref()
+            .unwrap()
+            .install(pf)
+            .map_err(|e| FilterError::InstallationError(e.to_string()))?;
+
+        self.filter_store
+            .as_ref()
+            .unwrap()
+            .add(f.clone())
+            .map_err(|e| {
+                self.tipset_filter_manager
+                    .as_ref()
+                    .unwrap()
+                    .remove(&f.id())
+                    .unwrap_or(());
+                FilterError::RemovalError(e.to_string())
+            })?;
+
+        Ok(f.id())
+    }
+}
+
+impl EthFilterSpec {
+    fn parse_eth_filter_spec(
+        &self,
+        chain_height: i64,
+        max_filter_height_range: i64,
+    ) -> Result<ParsedFilter, String> {
+        let mut min_height = 0;
+        let mut max_height = 0;
+        let mut tipset_cid = Cid::default();
+        let mut addresses = Vec::new();
+        if let Some(block_hash) = &self.block_hash {
+            if self.from_block.is_some() || self.to_block.is_some() {
+                return Err("must not specify block hash and from/to block".to_string());
+            }
+
+            tipset_cid = Cid::try_from(block_hash.0.as_bytes()).map_err(|e| e.to_string())?;
+        } else {
+            let (min, max) = parse_block_range(
+                chain_height,
+                self.from_block.as_deref(),
+                self.to_block.as_deref(),
+                max_filter_height_range,
+            )?;
+            min_height = min;
+            max_height = max;
+        }
+
+        for ea in &self.address {
+            let addr = ea
+                .to_filecoin_address()
+                .map_err(|e| format!("invalid address {}", e))?;
+            addresses.push(addr);
+        }
+
+        let keys = parse_eth_topics(&self.topics)?;
+
+        Ok(ParsedFilter {
+            min_height,
+            max_height,
+            tipset_cid,
+            addresses,
+            keys: keys_to_keys_with_codec(keys),
+        })
+    }
+}
+
+fn parse_block_range(
+    heaviest: ChainEpoch,
+    from_block: Option<&str>,
+    to_block: Option<&str>,
+    max_range: ChainEpoch,
+) -> Result<(ChainEpoch, ChainEpoch), String> {
```

Comment:
Let's use `anyhow::Result` and not have `String` as error.

Reply:
fixed